keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+claim parameters themselves, drivers now manage resources and describe them
+using a specific "semantic model" pre-defined by Kubernetes. This allows
+components aware of this "semantic model" to make decisions about these
+resources without outsourcing them to some third-party controller. For example,
+the scheduler is now able to allocate claims rapidly, without back-and-forth
+communication with DRA drivers.
+
+At a high-level, this extension takes the following form:
+
+* DRA drivers publish their available resources in the form of a
+  `NodeResourceSlice` object on a node-by-node basis according to one of the
+  builtin "semantic models" known to Kubernetes. This object is stored in the
+  API server and available to the scheduler (or Cluster Autoscaler) to query
+  when a resource request comes in later on.
+
+* When a user wants to consume a resource, they create a `ResourceClaim`,
+  which, in turn, references a claim parameters object. This object defines how
+  many resources are needed and which capabilities they must have. Typically, it
+  is defined using a vendor-specific type which might also support configuration
+  parameters (i.e. parameters that are *not* needed for allocation but *are*
+  needed for configuration).
+
+* With such a claim in place, DRA drivers "resolve" the contents of any
+  vendor-specific claim parameters into a canonical form (i.e. a generic
+  `ResourceClaimParameters` object in the `resource.k8s.io` API group) which
+  the scheduler (or Cluster Autoscaler) can evaluate against the
+  `NodeResourceSlice` of any candidate nodes without knowing exactly what is
+  being requested. They then use this information to help decide which node to
+  schedule a pod on (as well as allocate resources from its `NodeResourceSlice`
+  in the process).
+
+* Once a node is chosen and a workload is sent to it, DRA drivers are
+  responsible for injecting any allocated resources into the Pod, according to
+  the resource choices made by the scheduler. This includes applying any
+  configuration information attached to the vendor-specific claim parameters
+  object used in the request.
+
+In this KEP, we define a single "semantic model" called
+`PartitionableResources`, and use it throughout the discussion to help describe
+the framework as a whole. This model is a natural extension to the API provided
+by the traditional device-plugin, which supports advertising resources as a
+finite set of opaque strings. The major difference being that under the new
+model, one can attach a list of attributes to those opaque strings which the
+scheduler can use to select a particular resource for allocation. Additionally,
+each resource can be divided into a set of pre-defined partitions, each with
+their own set of unique attributes attached to them. In this way, requests for
+resources can be satisfied either by a complete resource instance or by one of
+its partitions.  This model may get enhanced in the future and/or other models
```

Comment:
I would suggest adding a little note explaining that each model's semantics needs to be implemented in the different components (or with a library). The goal of adding this is to emphasize that we probably want a low-number of models, rather than a proliferation of many models.

Something like:
```
its partitions.

This model may get enhanced...get added as needed. Each model must have
clear semantics which can be implemented by any consuming component, such
as the scheduler.
```